File: Roman to integer/RTI.cpp
#include <iostream>
#include <unordered_map>
#include <string>

using namespace std;

class Solution {
public:
    // Function to convert Roman numeral to integer
    int romantoInt(string s) {
        unordered_map<char, int> roman = {{'I', 1}, {'V', 5}, {'X', 10}, {'L', 50}, {'C', 100}, {'D', 500}, {'M', 1000}};
        int total = 0;
        int prev = 0;

        // Traverse the Roman numeral in reverse order
        for (int i = s.length() - 1; i >= 0; i--) {
            int value = roman[s[i]];
            if (value < prev) {
                total -= value;
            } else {
                total += value;
            }
            prev = value;
        }
        return total;
    }

    // Function to convert integer to Roman numeral
    string inttoRoman(int num) {
        unordered_map<int, string> roman = {{1000, "M"}, {900, "CM"}, {500, "D"}, {400, "CD"},
                                             {100, "C"}, {90, "XC"}, {50, "L"}, {40, "XL"},
                                             {10, "X"}, {9, "IX"}, {5, "V"}, {4, "IV"}, {1, "I"}};

        string result = "";
        // Traverse through the Roman numeral values in descending order
        for (auto it = roman.begin(); it != roman.end(); ++it) {
            while (num >= it->first) {
                result += it->second;
                num -= it->first;
            }
        }
        return result;
    }
};

int main() {
    Solution sol;
    string ans = "y";

    while (ans == "y") {
        cout << "Menu\n1. Roman to Integer\n2. Integer to Roman\n";
        int choice;
        cout << "Enter your choice: ";
        cin >> choice;

        if (choice == 1) {
            string roman_number;
            cout << "Enter the Roman number: ";
            cin >> roman_number;
            int result = sol.romantoInt(roman_number);
            cout << "The integer value of " << roman_number << " is: " << result << endl;
        } else if (choice == 2) {
            int integer_number;
            cout << "Enter the integer: ";
            cin >> integer_number;
            string result = sol.inttoRoman(integer_number);
            cout << "The Roman numeral for " << integer_number << " is: " << result << endl;
        } else {
            cout << "Invalid choice!" << endl;
        }

        cout << "Quit [y/n]: ";
        cin >> ans;
    }

    return 0;
}

File: Roman to integer/RTI.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class RTI {

    // Function to convert Roman numeral to integer
    public int romanToInt(String s) {
        Map<Character, Integer> roman = new HashMap<>();
        roman.put('I', 1);
        roman.put('V', 5);
        roman.put('X', 10);
        roman.put('L', 50);
        roman.put('C', 100);
        roman.put('D', 500);
        roman.put('M', 1000);

        int total = 0;
        int prev = 0;

        // Traverse the Roman numeral in reverse order
        for (int i = s.length() - 1; i >= 0; i--) {
            int value = roman.get(s.charAt(i));
            if (value < prev) {
                total -= value;
            } else {
                total += value;
            }
            prev = value;
        }
        return total;
    }

    // Function to convert integer to Roman numeral
    public String intToRoman(int num) {
        Map<Integer, String> roman = new HashMap<>();
        roman.put(1000, "M");
        roman.put(900, "CM");
        roman.put(500, "D");
        roman.put(400, "CD");
        roman.put(100, "C");
        roman.put(90, "XC");
        roman.put(50, "L");
        roman.put(40, "XL");
        roman.put(10, "X");
        roman.put(9, "IX");
        roman.put(5, "V");
        roman.put(4, "IV");
        roman.put(1, "I");

        StringBuilder result = new StringBuilder();

        // Traverse through the Roman numeral values in descending order
        for (Map.Entry<Integer, String> entry : roman.entrySet()) {
            while (num >= entry.getKey()) {
                result.append(entry.getValue());
                num -= entry.getKey();
            }
        }
        return result.toString();
    }

    public static void main(String[] args) {
        RTI sol = new RTI();
        try (Scanner scanner = new Scanner(System.in)) {
            String ans = "y";

            while (ans.equals("y")) {
                System.out.println("Menu\n1. Roman to Integer\n2. Integer to Roman");
                System.out.print("Enter your choice: ");
                int choice = scanner.nextInt();

                switch (choice) {
                    case 1 -> {
                        System.out.print("Enter the Roman number: ");
                        String romanNumber = scanner.next();
                        int result = sol.romanToInt(romanNumber);
                        System.out.println("The integer value of " + romanNumber + " is: " + result);
                    }
                    case 2 -> {
                        System.out.print("Enter the integer: ");
                        int integerNumber = scanner.nextInt();
                        String resultRoman = sol.intToRoman(integerNumber);
                        System.out.println("The Roman numeral for " + integerNumber + " is: " + resultRoman);
                    }
                    default -> System.out.println("Invalid choice!");
                }

                System.out.print("Quit [y/n]: ");
                ans = scanner.next();
            }
        }
    }
}
